### PR TITLE
WIP: Shared-transformer + warmdown-aligned training (not final submis…

### PR DESCRIPTION
**Pull request by leofeasby**
Shared-transformer architecture with MLP×5 expansion, bigram hash embedding
  (4096),
  weight decay tuning, SWA, and warmdown-aligned training schedule.

  Key finding: performance is dominated by a late-stage phase transition driven
  by LR
  warmdown. Aligning WARMDOWN_ITERS with the wallclock budget moves this
  transition
  earlier and converts wasted steps into usable convergence.

  Results on 8×H100 (within 10-min budget):
  - ~1.26 bpb int8 roundtrip
  - ~1.30 bpb TTT LoRA

  Work in progress — not a final submission.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
# Sibling Draft Review for PR #420

**Date:** 2026-04-12  
**Reviewer:** Claude Agent (for Mato/@MatoTeziTanka)  
**NEEDS_LLM_REVIEW:** Yes  
**Status:** Ready for detailed code review

## PR Summary

**Title:** WIP: Shared-transformer + warmdown-aligned training (not final submis…  
**State:** OPEN  
**Mato blocking comment:** NO  

## Marker Analysis

| Marker | Found | Notes |
|--------|-------|-------|
| target_in_key | False | Custom loss key injection pattern |
| TTT | False | Test-Time Training integration |
| SLOT | False | Slot-based attention variant |
| custom_tokenizer | True | 6 patterns detected |

## Architecture Changes

**train_gpt.py modifications:** ~3287 lines changed  
**Quantization:** True  
**EMA:** True  
**LoRA:** True  
**GPTQ:** False

## Assessment

⚠ REQUIRES_REVIEW - Custom modifications detected:
- Custom tokenizer patterns found (6 instances)

## Recommendation

**REVIEW** — Detailed code inspection needed before merge

## Review Checklist

- [ ] Code style and formatting (PEP 8, consistency)
- [ ] Numerical stability of modifications
- [ ] Integration with existing pipeline
- [ ] Reproducibility (seed handling, RNG control)
- [ ] Performance metrics reported correctly
- [ ] No undocumented dependencies

## Next Steps

1. Full train_gpt.py code review
2. Validate loss computation and gradients
3. Check for integration issues with existing pipeline
4. Verify metrics reproducibility

---
**Generated for:** Mato (@MatoTeziTanka)  
**Scope:** NEEDS_LLM_REVIEW sweep  
**Action:** Draft only—no posts to GitHub


**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE-pending standard record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Classification via sibling-session agent (Haiku-backed). This review was drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by MatoTeziTanka** [comment]
# Sibling Draft Review for PR #420

**Date:** 2026-04-12  
**Reviewer:** Claude Agent (for Mato/@MatoTeziTanka)  
**NEEDS_LLM_REVIEW:** Yes  
**Status:** Ready for detailed code review

## PR Summary

**Title:** WIP: Shared-transformer + warmdown-aligned training (not final submis…  
**State:** OPEN  
**Mato blocking comment:** NO  

## Marker Analysis

| Marker | Found | Notes |
|--------|-------|-------|
| target_in_key | False | Custom loss key injection pattern |
| TTT | False | Test-Time Training integration |
| SLOT | False | Slot-based attention variant |
| custom_tokenizer | True | 6 patterns detected |

## Architecture Changes

**train_gpt.py modifications:** ~3287 lines changed  
**Quantization:** True  
**EMA:** True  
**LoRA:** True  
**GPTQ:** False

## Assessment

⚠ REQUIRES_REVIEW - Custom modifications detected:
- Custom tokenizer patterns found (6 instances)

## Recommendation

**REVIEW** — Detailed code inspection needed before merge

## Review Checklist

- [ ] Code style and formatting (PEP 8, consistency)
- [ ] Numerical stability of modifications
- [ ] Integration with existing pipeline
- [ ] Reproducibility (seed handling, RNG control)
- [ ] Performance metrics reported correctly
- [ ] No undocumented dependencies

## Next Steps

1. Full train_gpt.py code review
2. Validate loss computation and gradients
3. Check for integration issues with existing pipeline
4. Verify metrics reproducibility

---
**Generated for:** Mato (@MatoTeziTanka)  
**Scope:** NEEDS_LLM_REVIEW sweep  
**Action:** Draft only—no posts to GitHub


**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE-pending standard record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Classification via sibling-session agent (Haiku-backed). This review was drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by MatoTeziTanka** [comment]
### Community Review — WIP: Shared-transformer + warmdown-aligned training (not final submis…

**Compliance flag: N-gram family bug (target token in bigram+trigram hash keys)**


Head SHA: bc275be80c1808869b89238027aee5c91730d13a
PR: #420 (leofeasby — WIP: Shared-transformer + warmdown-aligned training)

---

## N-gram Family Bug — CLOSE TRIGGERED

Both bigram and trigram embeddings include the **target token** (`input_ids[t]`) in the hash key, which is the disqualifying bug.

**Bigram (line 872):**
```python
bigram_idx = (prev * 1619 + input_ids) % self.bigram_table_size
```
`prev[:, 1:] = input_ids[:, :-1]`, so `prev[t] = token[t-1]`. The hash is `hash(token[t-1], token[t])` — the current (target) token is in the key. This is data leakage: the embedding for position `t` encodes what the model is supposed to predict.

**Trigram (line 879):**
```python
trigram_idx = (prev2 * 1000003 + prev1 * 1619 + input_ids) % self.trigram_table_size
```
Same violation: `input_ids[t]` (the target) appears in the hash alongside two prior context tokens.

This is the exact N-gram family bug pattern (target token in lookup key). It is **not** the legal BigramHash variant (which hashes only prior context tokens). CLOSE is mandatory.

---

## TTT Check — CLEAN

The `eval_val_ttt_lora` function (lines 1030–1137) implements score-first-per-chunk correctly:

**Verdict:** CLOSE — N-gram family bug (target token in bigram+trigram hash keys).

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** Recommend **CLOSE** unless the author restructures the flagged code.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source. If this review misread your code, please call it out so I can re-audit manually.*
